libcontainer: remove all mount logic from nsexec

cyphar opened the pull request:
Split out from #3953 to be reviewed separately.

~~_Draft until #4124 is merged._~~

~~_Draft until #4003 is merged._~~

~~_Draft until #3967 is merged._~~

<hr>

With open_tree(OPEN_TREE_CLONE), it is possible to implement both the
id-mapped mounts and bind-mount source file descriptor logic entirely in
Go without requiring any complicated handling from nsexec.

This allows us to remove the amount of C code we have in nsexec, as well
as simplifying a whole host of places that were made more complicated
with the addition of id-mapped mounts and the bind sourcefd logic. The
one downside of this is that the bind sourcefd feature now depends on
Linux 5.4.

In addition, we can easily add support for id-mappings that don't match
the container's user namespace. The approach taken here is to use Go's
officially supported mechanism for spawning a process in a user
namespace, but (ab)use PTRACE_TRACEME to avoid actually having to exec a
different process. The most efficient way to implement this would be to
do clone() in cgo directly to run a function that just does
kill(getpid(), SIGSTOP) -- we can always switch to that if it turns out
this approach is too slow.

This also includes a fix for a fairly sneaky bug caused by commit 9c444070ec7b
("Open bind mount sources from the host userns") and further exacerbated by
commit fda12ab1014e ("Support idmap mounts on volumes"), where opening the
mount sources long before we do any of the container rootfs mounts results in
undesirable behaviour with regards to mount propagation and unconventional
mount entries that use the container filesystem as the source. The solution is
to have a thread that joins the container mount namespace and does on-demand
generation of mount fds when requested.

Fixes: https://github.com/opencontainers/runc/commit/fda12ab1014e0ee23323cff6ec27ce7809c1e531 ("Support idmap mounts on volumes")
Fixes: https://github.com/opencontainers/runc/commit/9c444070ec7bb83995dbc0185da68284da71c554 ("Open bind mount sources from the host userns")
Signed-off-by: Aleksa Sarai <cyphar@cyphar.com>

cyphar commented:
The test failure is due to `runc init` being stuck waiting for the `procRun` sync message to be sent (even though it has been sent):

```
goroutine 584 [syscall, 2 minutes]:
syscall.Syscall(0x20?, 0xc0000f76c8?, 0x800000?, 0x7ffff800000?)
	/opt/hostedtoolcache/go/1.21.0/x64/src/syscall/syscall_linux.go:69 +0x25
syscall.read(0xc0003d0ae0?, {0xc0003c0800?, 0x0?, 0x0?})
	/opt/hostedtoolcache/go/1.21.0/x64/src/syscall/zsyscall_linux_amd64.go:721 +0x38
syscall.Read(...)
	/opt/hostedtoolcache/go/1.21.0/x64/src/syscall/syscall_unix.go:181
internal/poll.ignoringEINTRIO(...)
	/opt/hostedtoolcache/go/1.21.0/x64/src/internal/poll/fd_unix.go:736
internal/poll.(*FD).Read(0xc0003d0ae0, {0xc0003c0800, 0x200, 0x200})
	/opt/hostedtoolcache/go/1.21.0/x64/src/internal/poll/fd_unix.go:160 +0x2ae
os.(*File).read(...)
	/opt/hostedtoolcache/go/1.21.0/x64/src/os/file_posix.go:29
os.(*File).Read(0xc0000a21f8, {0xc0003c0800?, 0xc0003c0600?, 0xc000052028?})
	/opt/hostedtoolcache/go/1.21.0/x64/src/os/file.go:118 +0x52
encoding/json.(*Decoder).refill(0xc0000de280)
	/opt/hostedtoolcache/go/1.21.0/x64/src/encoding/json/stream.go:165 +0x188
encoding/json.(*Decoder).readValue(0xc0000de280)
	/opt/hostedtoolcache/go/1.21.0/x64/src/encoding/json/stream.go:140 +0xa9
encoding/json.(*Decoder).Decode(0xc0000de280, {0x819ba0, 0xc0003f2540})
	/opt/hostedtoolcache/go/1.21.0/x64/src/encoding/json/stream.go:63 +0x75
github.com/opencontainers/runc/libcontainer.doReadSync(0xc0000a21f8)
	/home/runner/work/runc/runc/libcontainer/sync.go:104 +0x9c
github.com/opencontainers/runc/libcontainer.parseSync(0x96c600?, 0xc0000f7b68)
	/home/runner/work/runc/runc/libcontainer/sync.go:160 +0x45
github.com/opencontainers/runc/libcontainer.(*initProcess).start(0xc000074480)
	/home/runner/work/runc/runc/libcontainer/process_linux.go:581 +0xa15
github.com/opencontainers/runc/libcontainer.(*Container).start(0xc0002e00a0, 0xc0003de3c0)
	/home/runner/work/runc/runc/libcontainer/container_linux.go:331 +0x11a
github.com/opencontainers/runc/libcontainer.(*Container).Start(0xc0002e00a0, 0xc0003de3c0)
	/home/runner/work/runc/runc/libcontainer/container_linux.go:212 +0xdb
github.com/opencontainers/runc/libcontainer.(*Container).Run(0x975918?, 0xc0003de3c0)
	/home/runner/work/runc/runc/libcontainer/container_linux.go:225 +0x1d
github.com/opencontainers/runc/libcontainer/integration.TestBindMountAndUser(0xc0002101a0)
	/home/runner/work/runc/runc/libcontainer/integration/exec_test.go:1805 +0x625
testing.tRunner(0xc0002101a0, 0x8f4d50)
	/opt/hostedtoolcache/go/1.21.0/x64/src/testing/testing.go:1595 +0xff
created by testing.(*T).Run in goroutine 1
	/opt/hostedtoolcache/go/1.21.0/x64/src/testing/testing.go:1648 +0x3ad
```

This is very strange, looking into it...

cyphar commented:
This reminds me of #515, though obviously the issue is different. The read inside `json.NewDecoder` is blocking on a read even though the other side has definitely written the `procRun` sync message. I wonder if it would help if we used `sendmsg` and `recvmsg` directly (possibly with `SOCK_SEQPACKET`) -- we could probably also send the JSON and the file at the same time for synchronisation...

cyphar commented:
Fixing the blocking issue was pretty awful (and required adding a new synchronisation pipe so we can use `SOCK_SEQPACKET` for the standard synchronisation) but that works now. I might also want to change the synchronisation file passing to do everything in a single packet....

The new test failure is only on Ubuntu 20.04 because it's indirectly related to hybrid cgroups -- criu cannot checkpoint our `/sys/fs/cgroup/*` controller mounts because they have sharing rules they don't support. I'm a little confused how these changes could've broken that, because those mounts are still handled the "traditional" way. In addition, the failure appears to be some kind of race because it only fails for ~50% of the tests on Ubuntu 20.04...

```
     checkpoint_test.go:136: (00.007467) mnt: Inspecting sharing on 935 shared_id 0 master_id 26 (@./sys/fs/cgroup/blkio)
    checkpoint_test.go:136: (00.007470) Error (criu/mount.c:1088): mnt: Mount 935 ./sys/fs/cgroup/blkio (master_id: 26 shared_id: 0) has unreachable sharing. Try --enable-external-masters.
```

cyphar commented:
```
=== RUN   TestFdLeaks
    exec_test.go:1714: unexpected error: readdirent /proc/thread-self/fd: no such file or directory
--- FAIL: TestFdLeaks (0.31s)
```

I swear to god, these CI machines are actually fucking cursed.

cyphar commented:
The exceptionally ugly-to-debug checkpoint issues were seemingly caused by the Go scheduler scheduling the mount-setup thread on a thread which became the thread-leader and the mountinfo parser we use uses `/proc/self/mountinfo` (which gives information about the thread-group, not the calling thread). Switching that to `/proc/thread-self/mountinfo` resolved the issue, but I also went ahead and switched everything else to use `/proc/thread-self` where appropriate to try to avoid other issues popping up.

I don't understand why there was more than one error mode (one where there were no cgroup mounts in mountinfo, and one where the mounts were the container mounts) but since `/proc/thread-self/mountinfo` solved the issue it seems clear that was the underlying issue. There was probably some ugly race involved.

In my testing, this was purely a unit test issue because in an actual runc binary the scheduler doesn't end up scheduling things this way. Now there's another CI-only unit test issue related to procfs, which makes even less sense than the last...

cyphar commented:
Damn, the lack of `/proc/thread-self` on CentOS 7 means that we need to do more `LockOSThread`-ing all over the place. There are three options:

1. Just use `/proc/self` for the bits of `rootfs_linux.go` used during container setup, because in that case we aren't going to have the mount thread in our thread group and thus the need for `/proc/thread-self` goes away. However, some bits of `rootfs_linux.go` are used by the CRIU logic and thus I'm not entirely sure it's safe to use `/proc/self` for everything.
2. Rewrite some of the stuff in `rootfs_linux.go` such that we can contain the need for `LockOSThread` to just the mount code (this seems somewhat doable). Switching some of the code away from path-based logic would also allow us to avoid possible cases where GC could close files we are using through `/proc/self/fd/...`.
3. Screw it, just `LockOSThread` everywhere.

I'm gravitating towards 2 but if it turns out to be too difficult we can go with 1 instead...

The most frustrating thing is that all of these issues only show up in `go test` as far as I can tell. I don't know if the Go scheduler for `go test` has some `GORUINMYWEEK=yesplease` option enabled but it really does feel like it.

kolyshkin commented:
@cyphar two issues:

1. `squash! libcontainer: remove all mount logic from nsexec` probably shouldn be renamed (or merged into other commits).
2. Not sure what's wrong with Cirrus CI, but it doesn't want to clone the repo.

cyphar commented:
1. Will do, I wanted to make sure the rewrite-of-the-rewrite worked before doing the squash.
2. I pushed the branch to origin by accident and Cirrus has saved the origin branch name and won't re-fetch the PR (probably because the commit is the same). I think a force-push will fix it...

rata commented:
It seems @cyphar [is on vacations](https://github.com/opencontainers/runtime-spec/pull/1222#issuecomment-1694650460), does someone wants to picks this up? Would that be okay?

cyphar commented:
I will push the version that is rebased on top of #3967 (which fixes the test errors here). But if anything else is needed I can't really guarantee that I'll be able to get back to it before I get back in two weeks. Feel free to carry this if necessary @kolyshkin.

rata commented:
@cyphar can we carry your other open PRs forward too?

cyphar commented:
Yeah, of course.

rata commented:
@kolyshkin can you carry this forward, then?

lifubang commented:
It's so big that we need time to review and do some corner tests.

lifubang commented on libcontainer/cgroups/file.go:
```diff
@@ -147,7 +149,7 @@ func openFile(dir, file string, flags int) (*os.File, error) {
 		// to reopen cgroupFd and retry openat2.
 		fdStr := strconv.Itoa(cgroupFd)
-		fdDest, _ := os.Readlink("/proc/self/fd/" + fdStr)
+		fdDest, _ := os.Readlink(utils.ProcThreadSelf("fd/" + fdStr))
```
Why doesn't need `lockOSThread` here?

cyphar replied:
The readlink happens atomically, so we don't need to worry about being rescheduled and the original thread being killed.

cyphar replied:
It turns out we do need locking here for pre-3.17 kernels.

cyphar commented on libcontainer/apparmor/apparmor_linux.go:
```diff
@@ -26,12 +27,19 @@ func isEnabled() bool {
+
+	// Under AppArmor you can only change your own attr, so use
+	// /proc/thread-self/ instead of /proc/<tid>/ like libapparmor does.
+	attrPath := utils.ProcThreadSelf("attr/apparmor/" + attr)
```
We still need the locking for old kernels because using `/proc/self/task/<tid>` directly has the same issues in Go as thread-self.

I agree that a release function would be nicer, and maybe we can have a "raw" version for the handful of cases where thread locking is unnecessary.

cyphar replied:
Actually, maybe the best approach would be to do something like WithProcFd, where you pass a closure?

cyphar replied:
I went with the closer (returned defer func) style.

cyphar commented on libcontainer/cgroups/cgroups_test.go:
```diff
@@ -5,6 +5,8 @@ import (
 )
 
 func TestParseCgroups(t *testing.T) {
+	// We don't need to use /proc/thread-self here because runc always runs
```
Yeah, though the unfortunate thing is that knowing whether to use thread-self or not is a little bit delicate and so the comments (while repetitive) are kind of important.

cyphar commented:
Thanks @lifubang! Just be aware that while it would be nice to cherry-pick the thread-self patch, when I tried it became quite ugly to disentangle it from the mount changes.

lifubang commented:
Part I: #4003

lifubang commented:
> when I tried it became quite ugly to disentangle it from the mount changes.

After #4003 merged, I'll see whether it's worth to do or not.

kolyshkin commented:
Added "draft until #4003 is merged" to the description, and marked as draft.

lifubang commented:
And also need until #3967 merged.

kolyshkin commented:
@cyphar #4003 is merged; can you please rebase this one?

cyphar commented:
Now the other PRs have been merged, this is pretty much ready -- however, I am going to (as per @lifubang's suggestion) try to fix up the `ProcThreadSelf` API so that it's less footgun-prone. I'll mark it as ready tomorrow when I've cleaned that up. I really wish this wasn't necessary...

lifubang commented:
> I really wish this wasn't necessary...

Yes, it’s unnecessary at this time, it can be in a separate PR, let’s merge this ASAP and release 1.2.0-rc1

cyphar commented:
This is ready.

@lifubang I went with your suggestion of a closer callback. I tried to do it with closures and it was too ugly to use...

lifubang commented on vendor/github.com/moby/sys/user/lookup_unix.go:
```diff
@@ -149,9 +149,13 @@ func CurrentUserSubGIDs() ([]SubID, error) {
 }
 
 func CurrentProcessUIDMap() ([]IDMap, error) {
+	// We don't need to use /proc/thread-self here because uid_map is a
```
We should not modify files in the vendor dir, we should update it from `go.mod`.
But I wonder why the lint doesn't check this now?

cyphar replied:
Ah sorry, this was done automatically by the rebase, I forgot that we moved this to moby. I'll fix it up.

lifubang commented on libcontainer/mount_linux.go:
```diff
@@ -121,3 +179,75 @@ func syscallMode(i fs.FileMode) (o uint32) {
+		}
+	}
+
+	if m.IsIDMapped() {
```
How about move the code in this block to above(L214)? Then we don't need the next two condition checkes that should never happen.

cyphar replied:
I wrote it this way because it's clearer how it could be extended to support other kinds of mount configuration, but I guess it's better to write it the "obvious" way for now.

lifubang commented on libcontainer/mount_linux.go:
```diff
@@ -121,3 +179,75 @@ func syscallMode(i fs.FileMode) (o uint32) {
+			if m.Flags&unix.MS_REC == unix.MS_REC {
+				flags |= unix.AT_RECURSIVE
+			}
+			mountFd, err := unix.OpenTree(unix.AT_FDCWD, m.Source, flags)
```
How about rename the var name `mountFd` to `Fd`? Because its name is the same as the function name.

lifubang commented on libcontainer/mount_linux.go:
```diff
@@ -121,3 +179,75 @@ func syscallMode(i fs.FileMode) (o uint32) {
 	return
 }
+
+// mountFd creates an open_tree(2)-like mount fd from the provided
```
This function is not only used for id mapped mount. So we should change this description to more clearly.

cyphar replied:
I don't think it implies that it's only used for id-mapped mounts ("open_tree(2)-like mount fd from the provided configuration" sounds like it's a mount that is being configured and is a file descriptor representing a mount side a-la `open_tree(2)` but not necessary using `open_tree(2)`). But I can add a note that it is used for other kinds of mounts too.

kolyshkin commented on tests/integration/idmap.bats:
```diff
@@ -2,167 +2,329 @@
 
 load helpers
 
+OVERFLOW_UID="$(cat /proc/sys/kernel/overflowuid)"
+OVERFLOW_GID="$(cat /proc/sys/kernel/overflowgid)"
+
```
I am not sure it is well defined in bats when exactly this code is executed, so this kind of sticks out.

So maybe move it to `function setup()` for clarity (or, if you really want to be only executed once, to `helpers.bash`, which is, unlike this file, a bash script).

cyphar replied:
Fair enough.

kolyshkin commented on libcontainer/cgroups/v1_utils.go:
```diff
@@ -99,11 +101,29 @@ func tryDefaultPath(cgroupPath, subsystem string) string {
 func readCgroupMountinfo() ([]*mountinfo.Info, error) {
 	readMountinfoOnce.Do(func() {
-		cgroupMountinfo, readMountinfoErr = mountinfo.GetMounts(
+		// mountinfo.GetMounts uses /proc/self/mountinfo, however in runc we
+		// have threads that join different namespaces and thus the output of
+		// the thread-group's mountinfo can depend on Go scheduling. Instead,
+		// use thread-self as we know the current thread has the "correct"
+		// mount namespace.
+		//
+		// See <https://github.com/moby/sys/pull/128>.
```
It's a shame that PR is still not merged. Let's merge it and rely on updated mountinfo instead of adding this here.

kolyshkin replied:
@cyphar please vendor https://github.com/moby/sys/releases/tag/mountinfo%2Fv0.7.0

cyphar replied:
It seems that the `/proc/self/mountinfo` fallback (that we have in runc) is needed for our checkpoint tests -- #4109 is failing for this reason. I've opened moby/sys#135 to fix it.

cyphar replied:
Also, a `moby/sys` update is not a blocker for this PR -- I just mentioned it because with the right updates it wouldn't be necessary to use `ProcThreadSelf` here.

lifubang commented on libcontainer/rootfs_linux.go:
```diff
@@ -723,19 +730,23 @@ func checkProcMount(rootfs, dest, source string) error {
-		// the mount is a proc filesystem
-		if isproc {
+		// only allow a mount on-top of proc if its source is a procfs mount ...
+		if m.Device == "proc" {
```
I think maybe we should change to `isproc`, because in the code, the value of `m.Device` is always `bind` or other, but not `proc`.
https://github.com/opencontainers/runc/blob/22a7199bb7b9d1a92175b18c4db4fb46318f549b/libcontainer/rootfs_linux.go#L552-L553
https://github.com/opencontainers/runc/blob/22a7199bb7b9d1a92175b18c4db4fb46318f549b/libcontainer/rootfs_linux.go#L670-L671

kolyshkin reviewed (approved):
LGTM

lifubang commented on libcontainer/mount_linux.go:
```diff
@@ -121,3 +179,75 @@ func syscallMode(i fs.FileMode) (o uint32) {
+				return nil, fmt.Errorf("failed to create userns for %s id-mapping: %w", m.Source, err)
+			}
+			defer usernsFile.Close()
+			if err := unix.MountSetattr(int(mountFile.Fd()), "", unix.AT_EMPTY_PATH, &unix.MountAttr{
```
I think we should also set `AT_RECURSIVE` here if we use `rbind`?
So, maybe we should add a testcase for `rbind` with `idmapped mount`?

cyphar replied:
We haven't added support for `rbind` yet to runc (it still isn't in a released spec). I will do that in a future PR.

lifubang replied:
I got it now. But you have checked it when calling `unix.OpenTree`. Do you think we should remove this `AT_RECURSIVE` and add a warning log when calling `unix.OpenTree`?
https://github.com/opencontainers/runc/blob/ae3fd6b345cb6e7de7b0373211af3ddbc7862166/libcontainer/mount_linux.go#L215-L217

cyphar replied:
Yeah, I'll remove that -- I added it because we didn't have `rbind` at the time so it made the most sense to me to make `rbind` recursive.

cyphar replied:
Actually, I think it makes more sense to add `idmap` and `ridmap` support.

cyphar replied:
@lifubang That code is needed for `rbind` mounts to work. At the moment the implementation _does not_ recursively apply the id mapping configuration (`mount_setattr` is called without `AT_RECURSIVE`).

However, it's probably a better idea to just implement `ridmap` and `idmap` support.

lifubang replied:
Yes. But I’m curious about how to deal with ‘rbind,idmap’. Maybe we should treat it the same as ‘bind,idmap’?

cyphar replied:
`rbind` + `idmap` would mean you do an `rbind` but with only the top-level mount having the idmapping applied. `MS_BIND|MS_REC` is completely separate to `MOUNT_ATTR_IDMAP`. I already have patches adding support, I'll push them in a bit once I finish writing the tests.

lifubang commented on libcontainer/mount_linux.go:
```diff
@@ -121,3 +179,75 @@ func syscallMode(i fs.FileMode) (o uint32) {
+		// because it requires doing mounts inside a new anonymous mount
+		// namespace.
+		if m.IsIDMapped() {
+			flags := uint(unix.OPEN_TREE_CLONE | unix.O_CLOEXEC)
```
Although `unix.OPEN_TREE_CLOEXEC` == `unix.O_CLOEXEC`, I think it will be better use `unix.OPEN_TREE_CLOEXEC`.

cyphar replied:
Good catch.

lifubang commented:
Another question(I can't figure out why at this time):
In a container, we can't create an new file in a idmapped mount dir without userns, but we can do it in the host.
```
test@runc:/tmp/mount-1$ touch newfile
touch: cannot touch 'newfile': Value too large for defined data type
```

cyphar commented:
@lifubang The issue is that you are trying to create a file with an unmapped owner -- if your container is running as host uid 0 but the mapping doesn't have an entry for container uid 0, the kernel doesn't know what the owner should be (see `fsuidgid_has_mapping` the `-EOVERFLOW`s in `fs/namei.c` and other places).

> but we can do it in the host

How are you doing it on the host? Are you creating a similar id-mapped mount on the host and then creating a file as uid 0, or are you just creating a file on the underlying filesystem as uid 0?

cyphar commented:
Tests are failing because there's something wrong with the OBS repos...

lifubang commented:
> How are you doing it on the host? Are you creating a similar id-mapped mount on the host and then creating a file as uid 0, or are you just creating a file on the underlying filesystem as uid 0?

```bash
root@iZj6cbuphafi5wf5srplytZ:/home/test/ubuntutest# mkdir hostmount
root@iZj6cbuphafi5wf5srplytZ:/home/test/ubuntutest# /opt/mount-idmapped/mount-idmapped --map-mount=b:0:100000:1 --recursive /home/test/ubuntutest/source-1 /home/test/ubuntutest/hostmount
root@iZj6cbuphafi5wf5srplytZ:/home/test/ubuntutest# cat /etc/passwd | grep test
test:x:100000:100000:,,,:/home/test:/bin/bash
root@iZj6cbuphafi5wf5srplytZ:/home/test/ubuntutest# su test
test@iZj6cbuphafi5wf5srplytZ:~/ubuntutest$ echo test >> hostmount/test
test@iZj6cbuphafi5wf5srplytZ:~/ubuntutest$ ls hostmount/
sub  subro  test  test1  test2  testfrom100000  testfromroot
test@iZj6cbuphafi5wf5srplytZ:~/ubuntutest$ touch hostmount/newfile
test@iZj6cbuphafi5wf5srplytZ:~/ubuntutest$
```

lifubang commented:
> if your container is running as host uid 0 but the mapping doesn't have an entry for container uid 0, the kernel doesn't know what the owner should be

In the container, I exec into it with the user id 100000, the same as the host.

cyphar commented:
@lifubang The case you've described works without issues on my machine. What configuration are you using? Is it possible you swapped `hostId` and `containerId` in one of the configurations?

Using the following configuration for the mount (with the user set to `100` in `config.json`):

```json{
	"destination": "/host-tmp",
	"source": "/tmp/for-containers",
	"options": [
		"bind",
		"idmap"
	],
	"uidMappings": [
		{"hostId": 100, "containerId": 0, "size": 100}
	],
	"gidMappings": [
		{"hostId": 100, "containerId": 0, "size": 100}
	]
}
```

Everything works the way you would expect:

```
% mkdir -p /tmp/for-containers
% sudo mount-idmapped --map-mount b:0:100:100 /tmp/for-containers /tmp/for-containers-mapped
% sudo -u 100 touch /tmp/for-containers-mapped/from-host
% runc run -b bundle ctr
$ id
uid=100(_apt) gid=100(users) groups=100(users)
$ touch /host-tmp/from-container
$ exit
# ls -la /tmp/for-containers{,-mapped}
/tmp/for-containers:
total 0
drwxr-xr-x  2 root root   80 Nov 16 11:07 .
drwxrwxrwt 51 root root 1420 Nov 16 11:03 ..
-rw-r--r--  1 root root    0 Nov 16 11:01 from-container
-rw-r--r--  1 root root    0 Nov 16 11:01 from-host

/tmp/for-containers-mapped:
total 0
drwxr-xr-x  2 rtorrent users   80 Nov 16 11:07 .
drwxrwxrwt 51 root     root  1420 Nov 16 11:03 ..
-rw-r--r--  1 rtorrent users    0 Nov 16 11:01 from-container
-rw-r--r--  1 rtorrent users    0 Nov 16 11:01 from-host
```

lifubang commented:
> with the user set to `100` in `config.json`

Yes, if I set uid to 100, it works. But if I set uid to 0, and use ‘runc exec -t -u 100 ctr sh’, ‘touch newfile’ can’t work.

cyphar commented:
@lifubang You need to do `-u 100:100`. `-u 100` results in you running as uid 100 but gid 0 (and the gid is not mapped by the mount, so the operation fails). See:

```
% runc exec -t -u 100 ctr id
uid=100(_apt) gid=0(root) groups=0(root)
% runc exec -t -u 100 ctr touch /host-tmp/from-exec-uid-100
touch: cannot touch '/host-tmp/from-exec-uid-100': Value too large for defined data type
% runc exec -t -u 100:100 ctr touch /host-tmp/from-exec-uidgid-100 # works
```

(This is different to Docker's behaviour, where if you use a username you will get the user's configuration.)

cyphar commented:
Okay, I've added `idmap` and `ridmap` support, including the just-merged "use the userns configuration" behaviour. One particular thing to note is that I've implemented them as "extensions" (a-la `tmpcopyup`) because adding `MOUNT_ATTR_IDMAP` to `RecAttrs` feels like a really bad hack and requires working around other bits. (FWIW, I think embedding the representation of `unix.MountAttr` into our `state.json` representation was a bit of a mistake.)

lifubang commented on libcontainer/mount_linux.go:
```diff
@@ -121,3 +179,84 @@ func syscallMode(i fs.FileMode) (o uint32) {
+			// recursively. This allows you to create "rbind" mounts where only
+			// the top-level mount has an idmapping. I'm not sure why you'd
+			// want that, but still...
+			if m.Extensions&configs.EXT_REC_IDMAP != 0 {
```
A not important thing:
How about add one more condition check `m.Flags&unix.MS_REC == unix.MS_REC` besides this check?
For example: `"bind", "ridmap"`.

cyphar replied:
`bind,ridmap` is functionally the same as `bind,idmap`, so adding more logic doesn't really accomplish anything and IMHO the logic is much clearer if you only check for `MS_REC` for the `open_tree(2)` call (emulating `rbind`) and only `EXT_REC_IDMAP` for `mount_setattr(2)`, because it's clear that the two flags are not related.

lifubang commented on libcontainer/configs/validate/validator.go:
```diff
@@ -299,26 +299,18 @@ func checkIDMapMounts(config *configs.Config, m *configs.Mount) error {
-	if !filepath.IsAbs(m.Source) {
-		return fmt.Errorf("mount source not absolute")
+	if len(m.UIDMappings) == 0 || len(m.GIDMappings) == 0 {
+		if m.Extensions&configs.EXT_ANY_IDMAP != 0 {
```
I think we should not return two different errors. It should combine to one error like this:
`id-mapped mounts must have both uid and gid mappings specified or be used with a container using user namespaces`

cyphar replied:
Oops. I added this error before I made `EXT_*IDMAP` be auto-set for all id-mapped mounts,  but now the second error will never be returned.

(FWIW, I think there was a benefit for having two errors -- the "or be used with a container using user namespaces" is confusing if the user hasn't specified `idmap` or `ridmap` but has instead only specified one of the mappings, triggering the error.)

lifubang commented on libcontainer/specconv/spec_linux.go:
```diff
@@ -551,6 +561,21 @@ func createLibcontainerMount(cwd string, m specs.Mount) (*configs.Mount, error)
+		// container doesn't have a userns, the config validator will error
+		// this out.
+		if mnt.Extensions&configs.EXT_ANY_IDMAP != 0 && spec.Linux != nil {
+			mnt.UIDMappings = toConfigIDMap(spec.Linux.UIDMappings)
```
Maybe `user` namespace implementation in `runc` is always wrong? Because we may ignore a situation which is joining an existing user namespace. for example:
```js
{

          "type": "user",
          "path": "/proc/1/ns/user"
}
```
Please see:
https://github.com/opencontainers/runc/blob/32d433cf5284153dbf90af007f979f37592c50d6/libcontainer/container_linux.go#L1198-L1202
If we are joining an existing user ns, the value of `spec.Linux.*IDMappings` is required but will be ignored, so I think this logic is error.

If I'm right, I think you should read these mappings from the existing user ns path, and open an new issue to fix this user namespace implementation error.

cyphar replied:
Ah, you're right -- this logic is wrong. In that case we could just open the nspath and use it as the userns fd for `mount_setattr(2)`. I will have to re-think how the `state.json` stuff works though...

lifubang replied:
I have read documents about user ns, maybe user ns is a special ns, we should refuse to join an existing user ns in runc.
‘’’
It is not permitted to use setns() to reenter the caller's
              current user namespace.  This prevents a caller that has
              dropped capabilities from regaining those capabilities via
              a call to setns().
‘’’
From https://man7.org/linux/man-pages/man2/setns.2.html

lifubang replied:
Issue opened: #4122

cyphar replied:
Yeah, this is purely a bug in the validator and (after writing a test for this case with `idmap`, it's clear we don't have any integration tests validating that joining paths works).

rata commented:
@cyphar interesting, from the PR description

> In short, the issue is that because we do
> OPEN_TREE_CLONE on the host, the RootPropagation flag does not apply
> (nor any other mount propagation flags configured in config.json) and
> thus recursive bind-mounts can and will be leaked to the host.

Not sure I follow, this is an issue in current main? Joining the container mnt namespace [here][link] (as we do in the bind-mount sources [here](https://github.com/opencontainers/runc/blob/9fffadae83091e17a71e7fd733d8a1ea294cd7b7/libcontainer/nsenter/nsexec.c#L620)) would solve it?

I can't see how a detached mount is leaked into the host mnt ns, what am I missing? Do you have a config.json that does that with main handy?

[link]: https://github.com/opencontainers/runc/blob/9fffadae83091e17a71e7fd733d8a1ea294cd7b7/libcontainer/nsenter/nsexec.c#L675

cyphar commented:
@rata My first draft of this PR took the old C design and just rewrote it in Go, but some of the tests started failing and I eventually figured out it was related to mount propagation issues. After that, the issue with the current design became kind of obvious and I figured out that the on-demand setup is necessary.

While `setns` solves the most egregious issue, the core problem is that if you want to make sure the `mount` operation is perfectly identical to `mount` being called directly by `rootfs_linux.go`, you need to create the mountfds after all previous mount operations are done. This is actually a larger issue than just propagation (though that was the main issue I ran into in terms of test failures and thus happens to be what I mentioned when I first wrote the PR description) -- if you have mount entries that use the container rootfs as a source, previous mount entries would affect the mount source but we grab the mount source far too early in the current C design. The only solution is to do it on-demand (the way this PR does).

But yes, doing `setns` would fix the RootPropagation problem AFAICS. Bind-mounts in the current implementation happen to not be affected by the propagation issue because we open the path and use it through `/proc/self/fd/...` -- if we use `open_tree` then the `setns` would've meant it avoided that particular issue.

I can take a look at writing some tests for this behaviour.

> I can't see how a detached mount is leaked into the host mnt ns, what am I missing?

I don't remember if I ran a test with this (I debugged all of this ~2 months ago), but from memory the issue is this:

If you have an idmapped mount in a MS_SHARED mount from the host (despite the container having MS_PRIVATE or MS_SLAVE propagation set), the `open_tree` and `move_mount` will create an MS_SHARED mount that then ends up in the container whose peer group is that of the host mount. If you have a subsequent mount on top of that mount, it will be shared with the host. The issue isn't the detached mount itself, it's that it shares a peer group with the host despite the container having MS_PRIVATE or MS_SLAVE.

But, as I said above, the mount propagation problem isn't the largest issue -- it's that the effective mount order gets broken with the current implementation.

I can write an actual example tomorrow, it's quite late here atm.

rata commented:
@cyphar thanks a lot!

rata commented on tests/integration/mounts.bats:
```diff
@@ -96,14 +108,101 @@ function teardown() {
+	config_add_bind_mount "$PWD/bind-src/a" "/mnt/x/y/z"
+	# Create a recursive bind-mount that uses part of the current tree in the
+	# container.
+	config_add_bind_mount "$rootfs/mnt/x" "$rootfs/mnt/x/y/z/x" "${from_ctr_bind_args[@]}"
+	config_add_bind_mount "$rootfs/mnt/x/y" "$rootfs/mnt/x/y/z" "${from_ctr_bind_args[@]}"
```
Do we really support and want to support the source of the bind mount in the config.json to be inside the container rootfs? Or should we fail the validation if we detect this?

I mean, end users don't know the path for the container rootfs, Kubernetes doesn't know either... I really suspect this isn't used by users and I'm unsure we want to support it.

cyphar replied:
That is how things work with Kubernetes, but runc supports more than just Kubernetes. There is nothing in the runtime-spec that indicates this is not allowed, and I see no reason to now allow it.

Also, this test acts as a proxy for the root propagation issue, but it's easier to test this than the root propagation stuff directly.

rata replied:
I see no reason to allow it and unsure it can be a headache, especially if there is no one needing that. But sure, we can support it... :)

cyphar replied:
Which part of the runtime-spec says it's not permitted? :smile_cat:

In fact, given that relative paths to the bundle are permitted, you don't even need to know the complete path and thus is something you could reasonably do, though maybe not in Kubernetes necessarily:

> source (string, OPTIONAL) A device name, but can also be a file or directory name for bind mounts or a dummy. Path values for bind mounts are either absolute or relative to the bundle.

There are a few cases where I could imagine it being useful (wanting to create an rbind of a complicated mount tree without having to duplicate all of the mount entries again).

And also, as I mentioned in https://github.com/opencontainers/runc/pull/3985#discussion_r1414006869, these kinds of mounts already work with runc. It's just that they were broken when runc used the `nsexec.c`-based mount source fd code.

rata replied:
I didn't say it wasn't permitted. But sure, if they worked, makes sense to keep that :)

rata commented on tests/integration/mounts.bats:
```diff
@@ -96,14 +108,101 @@ function teardown() {
+	config_add_bind_mount "$PWD/bind-src/a" "/mnt/x/y/z"
+	# Create a recursive bind-mount that uses part of the current tree in the
+	# container.
+	config_add_bind_mount "$rootfs/mnt/x" "$rootfs/mnt/x/y/z/x" "${from_ctr_bind_args[@]}"
+	config_add_bind_mount "$rootfs/mnt/x/y" "$rootfs/mnt/x/y/z" "${from_ctr_bind_args[@]}"
```
What do we do in runc 1.1 with this bind-mount (I mean a bind-mount with a source inside the container rootfs)? Is the result exactly the same as now with this PR applied?

cyphar replied:
runc 1.1 includes 9c444070ec7b, so with user namespaces it has the same error as current `main`.

The first test (`runc run [mount order, container bind-mount source]`) indicates the correct behaviour (we don't do anything special for bind-mounts when not in a user namespace) and thus the key thing is whether we have the same behaviour in the cases where we need to get mountfds. These tests pass on this PR, and don't pass on `1.1.10` or `main` because of 9c444070ec7b.

rata replied:
Sure, for the bind mount sources case, but what for the other case (!bind mount source fds passing)? Are we keeping the same behavior as runc 1.1?

rata replied:
Perfect, that addresses my concern. Thanks!

rata commented:
Please ping when the conflicts here are solved and ready for review :)

rata commented on libcontainer/userns/usernsfd_linux.go:
```diff
@@ -0,0 +1,134 @@
+	hs.m.Lock()
+	defer hs.m.Unlock()
+
+	if hs.procs == nil {
+		hs.procs = make(map[string]*os.Process)
+	}
```
I'd just return an error if this happens, having one way to use this API seems better IMHO. I think all users call new before too, so it is safe to do it now probably. What do you think?

cyphar replied:
`new(Handles)` still results in `hs.procs` being `nil`. We need to initalise the map when `Get` is first called. The other option is to have a constructor function.

rata replied:
Ohh, I missed that. It's okay, then, sorry :)

rata commented on tests/integration/idmap.bats:
```diff
@@ -3,167 +3,329 @@
+
+function setup_idmap_single_mount() {
+	uidmap="$1"       # ctr:host:size
+	gidmap="${2:-$1}" # ctr:host:size
+	mountname="${3:-1}"
+	destname="${4:-$mountname}"
```
Isn't it simpler if we assign these unconditionally? (without the ":-" thingy)

cyphar replied:
It's easier to use if `destname` defaults to `mountname`. But the defaults for `gidmap` and `mountname` aren't actually used (when I first wrote these tests, they were, but after a few refactors they stopped being used).

rata commented on tests/integration/idmap.bats:
```diff
@@ -3,167 +3,329 @@
+	chown 201:222 source-multi2/bar.txt
+	chown 202:233 source-multi2/baz.txt
+
+	# 1000 uids, 500 gids
+	chown 5000528:6000491 source-multi3/foo.txt
+	chown 5000133:6000337 source-multi3/bar.txt
+	chown 5000999:6000444 source-multi3/baz.txt
```
What is this comment supposed to mean?

cyphar replied:
The owners are in a space of 1000 uids and 500 gids.

rata commented on libcontainer/rootfs_linux.go:
```diff
@@ -719,19 +730,23 @@ func checkProcMount(rootfs, dest, source string) error {
 			return nil
 		}
-		// only allow a mount on-top of proc if it's source is "proc"
-		isproc, err := isProc(source)
-		if err != nil {
-			return err
-		}
-		// pass if the mount is happening on top of /proc and the source of
-		// the mount is a proc filesystem
-		if isproc {
+		// only allow a mount on-top of proc if its source is a procfs mount ...
+		if m.Device == "proc" {
```
Why change the logic here in this commit?

cyphar replied:
1. The old check was incorrect (the source path is ignored for proper mounts of pseudo-filesystems like `tmpfs` or `sysfs`, so checking the filesystem type of the source only makes sense if it's a bind-mount).
2. We need to use `srcStatfs` because of the `/proc/thread-self` stuff.

I can make a separate commit fixing (1).

rata commented on libcontainer/utils/utils_unix.go:
```diff
@@ -98,3 +105,103 @@ func NewSockPair(name string) (parent, child *os.File, err error) {
+
+	err = fn(procfd)
+	// Make sure the fd stays alive until after the closure returns.
+	runtime.KeepAlive(fh)
```
Hmm, this doesn't seem nice. Why was this added? Was it being closed before `fh.Fd()` was used?

I've hit similar issues, but the solutions that I come up with doesn't seem to be possible here.

cyphar replied:
I didn't see it in practice, but the last user of `fh` is a while before we use `procfd` and thus AFAIU the Go GC is entitled to close the file if it happened to run between the two points. The "correct" way of doing this is to use `(*os.File).Sysconn` which guarantees the fd won't be freed but that interface is much uglier.

cyphar replied:
Oh actually, we do `defer fh.Close()` so it should be fine.

rata commented on libcontainer/utils/utils_unix.go:
```diff
@@ -98,3 +105,103 @@ func NewSockPair(name string) (parent, child *os.File, err error) {
+	})
+
+	// We need to lock our thread until the caller is done with the path string
+	// because any inatomic operation on the path (such as opening a file, then
```
Maybe?

```suggestion
	// because any nont-atomic operation on the path (such as opening a file, then
```

rata commented on libcontainer/utils/utils_unix.go:
```diff
@@ -98,3 +105,103 @@ func NewSockPair(name string) (parent, child *os.File, err error) {
+	// reading it) could be interrupted by the Go runtime where the underlying
+	// thread is swapped out and the original thread is killed, resulting in
+	// pull-your-hair-out-hard-to-debug issues in the caller. In addition, the
+	// pre-3.17 fallback makes everything inatomic because the same thing could
```
idem

rata reviewed (commented):
Left some comments, this mostly LGTM.

The errors generated are decent and I tested this with containerd and Kubernetes, with >6.3, 6.1 and 5.10 kernels (so kernel with idmap mount support for all the fs, kernel with idmap mount support to _some_ fs, and kernel without the syscalls implemented).

However, for this to work with containerd I had to do this change:
```
$ git diff
diff --git libcontainer/specconv/spec_linux.go libcontainer/specconv/spec_linux.go
index 991962c4..203bf694 100644
--- libcontainer/specconv/spec_linux.go
+++ libcontainer/specconv/spec_linux.go
@@ -1006,9 +1006,9 @@ func setupUserNamespace(spec *specs.Spec, config *configs.Config) error {
        if path := config.Namespaces.PathOf(configs.NEWUSER); path != "" {
                // We cannot allow uid or gid mappings to be set if we are also asked
                // to join a userns.
-               if config.UIDMappings != nil || config.GIDMappings != nil {
-                       return errors.New("user namespaces enabled, but both namespace path and mapping specified -- you may only provide one")
-               }
+               //if config.UIDMappings != nil || config.GIDMappings != nil {
+               //      return errors.New("user namespaces enabled, but both namespace path and mapping specified -- you may only provide one")
+               //}
                // Cache the current userns mappings in our configuration, so that we
                // can calculate uid and gid mappings within runc. These mappings are
                // never used for configuring the container if the path is set.
```

Those lines are not part of this PR, though.

CRIO triggers the very same error too.

I think we should change that to a warning, change CRIO, containerd (and maybe more tools), and change it back to an error in a few releases. I haven't checked what is done today.

I think both (containerd and CRIO) when sending both, the path and the mappings, those are consistent (i.e. the path's mappings and the mappings in the config.json are the same). In that case, we can just print a warning here saying we will ignore one (probably the mappings and just use the path) would be safe.

lifubang commented:
> I think we should change that to a warning, change CRIO, containerd (and maybe more tools), and change it back to an error in a few releases. I haven't checked what is done today.

I have thought it before in #4124, maybe we can also combine them to an new (u|g)idMappings?, I think this is not clarified in `runtime-spec` when both user ns path and (u|g)idMappings are provided.
How about open an new issue to record this? Because this PR has been big enough.

cyphar commented:
Configurations that have userns mappings and a userns path to join specified are fundamentally broken. As a stop-gap, I can make a **separate** PR which will output a warning instead of an error _if and only if_ the mappings are identical.

But fundamentally such configurations don't make sense because you cannot reconfigure a userns that has already has it's mappings configured. (In theory you could create a new userns but not apply any mappings, and then hand that userns to runc. But that's not what's going on with containerd. And I'm not sure such a setup would work anyway.)

rata commented:
@cyphar 
> As a stop-gap, I can make a **separate** PR which will output a warning instead of an error _if and only if_ the mappings are identical.

That should work. Please cc me so I can verify it works fine with containerd and crio.

cyphar commented on libcontainer/mount_linux.go:
```diff
@@ -121,3 +167,75 @@ func syscallMode(i fs.FileMode) (o uint32) {
+		// We only need to use OPEN_TREE_CLONE in the case where we need to use
+		// mount_setattr(2). We are currently in the container namespace and
+		// thus there is no risk of an opened directory being used to escape
+		// the container. OPEN_TREE_CLONE is more expensive than open(2)
+		// because it requires doing mounts inside a new anonymous mount
+		// namespace.
```
Yes, because that's the reason we're not using `OPEN_TREE_CLONE` for everything (I mean, kernel support is another reason but we could have a fallback if we really wanted to use `OPEN_TREE_CLONE` when available).

rata replied:
Oh, okay, then can you mention that in the comment? My thought was really: "do we care? Do we care enough to not use it unconditionally? I thought the kernel support was the reason". If we state what you said, that would at least avoid me doing that simplification in the future :-D

cyphar replied:
I mean, that _is_ exactly what the comment is explaining, but I can try to make it a bit clearer.

cyphar commented on libcontainer/mount_linux.go:
```diff
@@ -121,3 +167,75 @@ func syscallMode(i fs.FileMode) (o uint32) {
+		// because it requires doing mounts inside a new anonymous mount
+		// namespace.
+		if m.IsIDMapped() {
+			flags := uint(unix.OPEN_TREE_CLONE | unix.OPEN_TREE_CLOEXEC)
```
I'm not sure whether we want to block auto-mounting (we don't do that anywhere else), and we follow symlinks when doing regular bind-mounts so I'm not sure whether we want to block symlinks here. And we can't block symlinks for block devices because stuff like `/dev/mapper/...` and `/dev/by-id/` are all symlinks.

I get the concern, but `config.json` is a trusted configuration source (this is runc's threat model -- I know Kubernetes feels differently on the topic, but they're wrong to think so), and IMHO it's more important to have consistent behaviour for all of our mount types.

cyphar commented on libcontainer/process_linux.go:
```diff
@@ -487,6 +590,19 @@ func (p *initProcess) start() (retErr error) {
 
+	// Spin up a goroutine to handle remapping mount requests by runc init.
+	// There is no point doing this for rootless containers because they spawn
+	// the idmap processes, nor can they configure MOUNT_ATTR_IDMAP.
```
In pure rootless, it's not possible because both `open_tree(OPEN_TREE_CLONE)` and `mount_setattr` require `may_mount()` (`ns_capable(current->nsproxy->mnt_ns->user_ns, CAP_SYS_ADMIN)`) which we don't have with rootless containers. We could pass plain `open(2)` paths for bind-mounts but there's no point because in that case there is no permission issue with bind-mounts (you cannot `setgroups(2)` in rootless and you never change kuid/kgid).

cyphar commented on libcontainer/mount_linux.go:
```diff
@@ -121,3 +167,75 @@ func syscallMode(i fs.FileMode) (o uint32) {
+	if !m.IsBind() {
+		return nil, errors.New("new mount api: only bind-mounts are supported")
+	}
+	if nsHandles == nil {
+		nsHandles = new(userns.Handles)
+		defer nsHandles.Release()
+	}
```
Yes, but there's no way of knowing that at compile-time, so we'd need a runtime check anyway -- and it seems nicer to service the request if `userns.Handles` wasn't provided than return an error for a programming bug.

rata commented on libcontainer/rootfs_linux.go:
```diff
@@ -733,14 +733,20 @@ func checkProcMount(rootfs, dest string, m *mountEntry) error {
-			return err
-		}
-		if st.Type == unix.PROC_SUPER_MAGIC {
+		// only allow a mount on-top of proc if its source is a procfs mount ...
+		if m.Device == "proc" {
```
What I'd like is the commit message to explain why mounting procfs on top of procfs is safe. 

The commit seems confusing also, as it says it's better to be safe than sorry, but it allows more cases than before.

cyphar replied:
Previously any mount that had a source path that was on procfs (even if it wasn't a bind-mount) was allowed because there were no checks that it was a bind-mount. This adds a check for that, and is what I'm referring to as being "better safe than sorry".

For what it's worth, the `m.Device == "proc"` is never hit anyway, so I guess we can drop it. But it is obviously safe, I don't see how being able to mount procfs on top of `/proc` is unsafe...

rata replied:
I see. I wonder if we are making some mounts fail now and its worth a changelog line. IIUC a mount with a source that was a proc fs, without the "bind" option, worked before. The fix here is only requiring it to contain the option "bind".

Is this correct?

And yeah, if the m.Device == proc case can't happen, please update the code and commit message.

cyphar replied:
Taking another look, we should be calling `checkProcMount` in more places. Adding the check to all mount types (which is correct) means that we need to check `m.Device == "proc"`.

rata replied:
Uhm, this commit is changing a lot in the last few hours.

How would an example of this be look like in the config.json?

> This meant that having a non bind-mount entry for a pseudo-filesystem (like
overlayfs) with a dummy source set to /proc on the host would let you
bypass the check,

Any reason to not add a test for this?

Or even better, should this be a separate PR, with its own explanation, tests and be backported to 1.1?

rata replied:
But if that is too much, just let me know when the commit/code settles :)

cyphar replied:
There was only one new version since you last looked at it -- I added extra `checkProcMount` checks which were missing, and I moved the `m.Device == "proc"` check to the start of the function. The reasons for this change are pretty self-explanatory.

`checkProcMount` protects against mounts like 

```json
{
    "destination": "/proc",
    "type": "bind",
    "source": "/my/fake/proc",
    "options": [
        "bind"
    ]
}
```

The bug was that mounts like

```
{
    "destination": "/proc",
    "type": "overlay",
    "source": "/proc",
    "options": [
        "lowerdir=/tmp/proc1",
        "upperdir=/tmp/proc2",
        "workdir=/tmp/work"
    ]
}
```

were not blocked with the old code (note that the source path is ignored by overlayfs). 

Yes, this is a potential security issue. No, I don't think it's CVE-worthy because untrusted users shouldn't have the ability to configure arbitrary mounts (especially proper filesystem mounts, as opposed to stuff like tmpfs -- `tmpfs` is also not protected against until this commit, but `tmpcopyup` doesn't work so it's not a suitable attack method).

For what it's worth, I think we should be more strict than this. We should be checking that the inode of the mount source is equal to `PROC_ROOT_INO` (== 1). But I'm not sure if that will break real users.

There is an existing test which depends on this, in fact:

```go
func TestCheckMountDestOnProcChroot(t *testing.T) {
	m := mountEntry{
		Mount: &configs.Mount{
			Destination: "/proc",
			Source:      "foo",
			Device:      "proc",
		},
	}
	dest := "/rootfs/proc/"
	err := checkProcMount("/rootfs", dest, &m)
	if err != nil {
		t.Fatalf("destination inside proc when using chroot should not return an error: %v", err)
	}
}
```

> Any reason to not add a test for this?

I didn't think it was really that necessary, but I can add some unit tests to `libcontainer/rootfs_linux_test.go`.

> Or even better, should this be a separate PR, with its own explanation, tests and be backported to 1.1? But if that is too much, just let me know when the commit/code settles :)

I've spent several months on this PR now, I don't see a strong reason to do this. It's not as simple as a cherry-pick -- this PR changes several other aspects of `rootfs_linux.go` (including the signature of `checkProcMount` in a way that would be annoying to split and then rebase on top of) and some of them are required for this change. And I'm not sure if we will have a 1.1.11 release, given that 1.2.x is almost done once this is merged.

rata replied:
SGTM, thanks

rata commented on libcontainer/rootfs_linux.go:
```diff
@@ -733,14 +733,20 @@ func checkProcMount(rootfs, dest string, m *mountEntry) error {
+				return nil
+			}
+		}
 		return fmt.Errorf("%q cannot be mounted because it is not of type proc", dest)
```
It seems this line might not be accurate with this commit now, as the source could be of type proc and just not contain the "bind" option.

I'm unsure if we are improving things with this commit, it's not obvious at least.

cyphar replied:
"is it not of type proc", is correct in both cases....

cyphar replied:
To be clear, the issue I am fixing here is that `checkProcMount` doesn't currently return an error if you add (for example) an `overlay` mount that has a "source" of `/proc` but actually has a different `lowerdir`.

Hopefully that makes it more obvious that this clearly needs to be fixed.

rata commented on libcontainer/rootfs_linux_test.go:
```diff
@@ -52,7 +51,41 @@ func TestCheckBindMountOnProc(t *testing.T) {
+	}
+}
+
+func TestCheckTrickyMountOnProc(t *testing.T) {
```
These two seem c&p and table tests might be a better way to avoid the duplication here. It can be on a follow-up PR if you prefer that too.

cyphar replied:
Yeah, I just wanted to keep it as-is for now.

rata commented on libcontainer/rootfs_linux.go:
```diff
@@ -747,12 +740,28 @@ func checkProcMount(rootfs, dest string, m mountEntry) error {
-			return err
-		}
-		if st.Type == unix.PROC_SUPER_MAGIC {
+		// Onlu allow bind-mounts on top of /proc, and only if the source is a
```
```suggestion
		// Only allow bind-mounts on top of /proc, and only if the source is a
```

rata commented on libcontainer/rootfs_linux.go:
```diff
@@ -747,12 +740,28 @@ func checkProcMount(rootfs, dest string, m mountEntry) error {
-		if st.Type == unix.PROC_SUPER_MAGIC {
+		// Onlu allow bind-mounts on top of /proc, and only if the source is a
+		// procfs mount.
+		if m.IsBind() {
```
nit: I think this looks cleaner if we first return on the cases we are not interested on. Like: `if m.Device == "proc" && !m.IsBind()...; else if !m.IsBind()` and then just continue with this without identation.

cyphar replied:
My worry was that it's very easy to miss that you need to be careful about how you check the type and `m.IsBind`. Also, the structure makes it clearer that it's fail-closed (this is how the old version worked too, and I didn't feel it was necessary to change it).

rata commented on libcontainer/mount_linux.go:
```diff
@@ -229,16 +229,32 @@ func mountFd(nsHandles *userns.Handles, m *configs.Mount) (*mountSource, error)
+			usernsFile, err = os.Open(m.Mapping.NamespacePath)
+			if err != nil {
+				return nil, fmt.Errorf("failed to open existing userns for %s id-mapping: %w", m.Source, err)
+			}
```
I think the path joining logic is on this commit by mistake, but it is fine.

cyphar replied:
Which path joining logic?

rata replied:
Never mind.

This code is joining an existing namespace with `os.Open()`. However, this is part of the commit `mount: add support for ridmap and idmap`, which is confusing.

I think you might have done this to avoid conflicts and writing the same func again, as here you switch to m.Mapping.NamespacePath. And I agree it is needed. A mention that we are also doing the joining logic now in the commit would have helped me to review, as this is unexpected after reading the commit.

But never mind.

cyphar replied:
The reason this was added here is because only `idmap` and `ridmap` allow you to use the container's user namespace configuration (specifically, the userns path) as the source for the idmap attribute. Before that commit, every mount needed to specify the mappings explicitly.

rata replied:
Ohhh, sorry I missed it. Thanks!

rata commented on libcontainer/rootfs_linux.go:
```diff
@@ -1299,6 +1299,12 @@ func setRecAttr(m *configs.Mount, rootfs string) error {
+	// Make sure that we don't have MOUNT_ATTR_IDMAP specified here. id-mapped
+	// mounts are handled through procMountPlease, and we cannot apply this
+	// attribute here. This should never happen.
+	if (m.RecAttr.Attr_set|m.RecAttr.Attr_clr)&unix.MOUNT_ATTR_IDMAP != 0 {
+		return errors.New("[internal error] mount configuration cannot contain recAttr for MOUNT_ATTR_IDMAP")
+	}
```
Do we want to validate this on the validator too? :thinking:

cyphar replied:
Good point. I put this here when the way `idmap` and `ridmap` were represented were such that you couldn't do the check in the validator.

rata reviewed (commented):
This mostly LGTM, I left several questions and some small comments. Thanks!

btw, I haven't reviewed the `tests/integration/` folder. My personal preference is to not call so many functions with side effects in integration tests (so it VERY simple to have a local repro without running the tests), and I prefer some code duplication even if that helps to make it really obvious what is happening. But that is just my taste :)

cyphar commented on libcontainer/rootfs_linux.go:
```diff
@@ -62,43 +64,64 @@ func needsSetupDev(config *configs.Config) bool {
+		// necessary. For bind-mounts, this is only necessary if we cannot
+		// resolve the parent mount (this is only hit if you are running in a
+		// userns -- but for rootless the host-side thread can't help).
+		wantSourceFile := m.IsIDMapped()
+		if m.IsBind() && !config.RootlessEUID {
+			if _, err := os.Stat(m.Source); err != nil {
+				wantSourceFile = true
+			}
```
> Is this right? Shouldn't wantSourceFile be true if is idmap OR if is bind, non-rootless and can't stat?

That is exactly what this code does.

rata replied:
haha, brain fart. I don't know what I was thinking yesterday to arrive to another conclusion. Sorry :D

cyphar commented on libcontainer/rootfs_linux.go:
```diff
@@ -718,18 +747,12 @@ func checkProcMount(rootfs, dest, source string) error {
-		// pass if the mount is happening on top of /proc and the source of
-		// the mount is a proc filesystem
-		if isproc {
+		if st.Type == unix.PROC_SUPER_MAGIC {
```
Yes, we need to change this code to use `srcStatfs()` as part of this commit, and there are no other users of `isProc` so removing it makes the code easier to read. Otherwise we'd need to change the signature of a function we remove in a later commit.

cyphar commented on libcontainer/rootfs_linux.go:
```diff
@@ -732,6 +720,11 @@ func getCgroupMounts(m *configs.Mount) ([]*configs.Mount, error) {
 	return binds, nil
 }
 
+// Taken from <include/linux/proc_ns.h>. If a file is on PROC_SUPER_MAGIC,
```
That's not what this is saying, it's saying if a file is on a *filesystem of type* PROC_SUPER_MAGIC. But I can write that instead.

cyphar commented on libcontainer/apparmor/apparmor_linux.go:
```diff
@@ -26,14 +26,19 @@ func isEnabled() bool {
-	attrPath := "/proc/self/attr/apparmor/" + attr
-	if _, err := os.Stat(attrPath); errors.Is(err, os.ErrNotExist) {
+	attr = utils.CleanPath(attr)
+	attrSubPath := "attr/apparmor/" + attr
+	if _, err := os.Stat("/proc/self/" + attrSubPath); errors.Is(err, os.ErrNotExist) {
 		// fall back to the old convention
-		attrPath = "/proc/self/attr/" + attr
+		attrSubPath = "attr/" + attr
 	}
 
+	// Under AppArmor you can only change your own attr, so there's no reason
+	// to not use /proc/thread-self/ (instead of /proc/<tid>/, like libapparmor
+	// does).
```
I'm not sure what you mean by "do the opposite"? We still don't use `/proc/<tid>`.

(To be honest, I'm not sure this behaviour even needs a comment because it is obviously correct, but I've kept it from when we switched away from libapparmor several years ago.)

rata replied:
SGTM.

I meant the comment before says: don't use proc/<tid>
Now it says: there is no reason to not use proc/thread-self/.

I think yesterday I was having issues with a "logic not", this SGTM :)

cyphar commented on libcontainer/mount_linux.go:
```diff
@@ -101,19 +102,30 @@ func mountViaFds(source string, srcFile *mountSource, target, dstFd, fstype stri
+	isMoveMount := srcFile != nil && srcFile.Type == mountSourceOpenTree
 	if srcFile != nil {
-		src = "/proc/self/fd/" + strconv.Itoa(int(srcFile.file.Fd()))
+		// If we're going to use the /proc/thread-self/... path for classic
+		// mount(2), we need to get a safe handle to /proc/thread-self. This
+		// isn't needed for move_mount(2) because in that case the path is just
+		// a dummy string used for error info.
+		fdStr := strconv.Itoa(int(srcFile.file.Fd()))
+		if isMoveMount {
+			src = "/proc/self/fd/" + fdStr
```
We don't need to `runtime.LockOSThread` because in the `move_mount(2)` case we don't use the source path string and so the string is only used for error messages (the comment mentions this).

rata reviewed (approved):
LGTM

cyphar commented:
/ping @AkihiroSuda @lifubang :pray:

AkihiroSuda commented on libcontainer/configs/mount_linux.go:
```diff
@@ -2,6 +2,24 @@ package configs
+	// NamespacePath is a path to a user namespace that indicates the necessary
+	// id-mappings for MOUNT_ATTR_IDMAP. If set to non-"", UIDMappings and
+	// GIDMappings must be set to nil.
+	NamespacePath string `json:"userns_path,omitempty"`
```
```suggestion
	NamespacePath string `json:"namespace_path,omitempty"`
```
(Or rename the Go variable to match the json field name)

AkihiroSuda commented on libcontainer/configs/mount_linux.go:
```diff
@@ -34,23 +52,15 @@ type Mount struct {
-	GIDMappings []IDMap `json:"gid_mappings,omitempty"`
+	// Mapping is the MOUNT_ATTR_IDMAP configuration for the mount. If non-nil,
+	// the mount is configured to use MOUNT_ATTR_IDMAP-style id mappings.
+	Mapping *MountMapping `json:"id_mapping,omitempty"`
```
The go name does not match the json name

AkihiroSuda commented on libcontainer/configs/validate/validator_test.go:
```diff
@@ -504,103 +504,108 @@ func TestValidateIDMapMounts(t *testing.T) {
 		config *configs.Config
 	}{
 		{
-			name:  "idmap mount without bind opt specified",
```
Can we retain existing tests?

cyphar replied:
I haven't removed any tests. The diff makes it look weird, but this just changes the name of the test, and makes it actually test a realistic configuration (it has a valid `Device` value).

AkihiroSuda commented on libcontainer/process_linux.go:
```diff
@@ -398,6 +404,106 @@ func (p *initProcess) waitForChildExit(childPid int) error {
+		defer nsHandles.Release()
+	loop:
+		for {
+			select {
```
Don’t we need some timeout?

cyphar replied:
We have the cancel function for the context, which is deferred by the only caller, and so this will always exit once the caller returns. I guess we could add a `WithDeadline` for the context, but making it possible for the thread to go away randomly seems like it could cause panics or other deadlocks on channels.

cyphar replied:
I've reworked the code and gone with a 1-minute timeout. That should be long enough.

AkihiroSuda commented on libcontainer/userns/usernsfd_linux.go:
```diff
@@ -0,0 +1,156 @@
+	// We don't need to use /proc/thread-self here because the exe mm of a
+	// thread-group is guaranteed to be the same for all threads by definition.
+	// This lets us avoid having to do runtime.LockOSThread.
+	return os.StartProcess("/proc/self/exe", []string{"runc", "--help"}, &os.ProcAttr{
```
This can’t be just /bin/true ? (when it exists)

cyphar replied:
The path doesn't actually matter, we never call `execve(2)`. We can't use `""` or some other dummy path because the Go runtime checks that the path exists. I would prefer to not add code to spawn a random external executable (even if we don't actually execute it), and if `/bin/true` doesn't exist we would need a fallback anyway.

AkihiroSuda reviewed (approved):
LGTM, but please consider to keep json field names consistent with Go variable names

cyphar commented:
@AkihiroSuda I'm fixing them up now :smile_cat:.

AkihiroSuda commented on libcontainer/mount_linux.go:
```diff
@@ -1,22 +1,48 @@
+)
+
+type mountSource struct {
+	Type mountSourceType `json:"source_type"`
```
nit: name mismatch

cyphar commented:
@kolyshkin Did you want to re-review this, or can we merge it as-is?

lifubang commented on libcontainer/mount_linux.go:
```diff
@@ -121,3 +179,97 @@ func syscallMode(i fs.FileMode) (o uint32) {
+			if err != nil {
+				return nil, fmt.Errorf("failed to create userns for %s id-mapping: %w", m.Source, err)
+			}
+		} else {
```
It seems like there is no tests for this condition?

cyphar replied:
The last test in `idmap.bats` (`idmap mount (idmap flag, implied mapping, userns join) [userns]`) should test this, unless I'm missing something?

lifubang replied:
oh, yeah, let's move forward.

lifubang commented:
https://github.com/opencontainers/runc/blob/fa93c8b05b323187d0c5e3bdb3622c23c3abcfff/libcontainer/rootfs_linux.go#L734
1. Maybe here it should be: If m.Source is nil......
2. And maybe this line can be removed, because it seems like we don't care about it anymore.

cyphar commented:
> https://github.com/opencontainers/runc/blob/fa93c8b05b323187d0c5e3bdb3622c23c3abcfff/libcontainer/rootfs_linux.go#L734
> 1. Maybe here it should be: If m.Source is nil......
> 2. And maybe this line can be removed, because it seems like we don't care about it anymore.

Yeah, that comment was outdated by the change we backported here. (I didn't notice until this backpott that I removed this weird criu restore behaviour by fixing the proc check.)

cyphar commented:
@lifubang Yeah, that comment was obsoleted by cdff09ab875159d004035990c0d45e8bdf20ed35 ("rootfs; fix 'can we mount on top of /proc' check") but I didn't notice this until I did the backport in #4334.

(We can remove it, hopefully this didn't break anything.)
